src/Security/InheritedPermissions.php

```diff
@@ -0,0 +1,632 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use InvalidArgumentException;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\Hierarchy\Hierarchy;
+use SilverStripe\Versioned\Versioned;
+
+/**
+ * Calculates batch permissions for nested objects for:
+ *  - canView: Supports 'Anyone' type
+ *  - canEdit
+ *  - canDelete: Includes special logic for ensuring parent objects can only be deleted if their children can
+ *    be deleted also.
+ */
+class InheritedPermissions
```

Comment:
How about PermissionChecker interface (implemented by InheritedPermissionChecker) and ScopedPermissionChecker (inherited by the root permission classes, each of which will be renamed)

the difference being ScopedPermissionChecker doesn't take any IDs